### DISCUSSION TITLE
Welcome to LangChain Discussions!

### DISCUSSION THREAD

**Discussion by HavenDV** [tryAGI/LangChain#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by reinarrow** [comment]
Hi there @HavenDV and others. I'm a passionate programmer making some huge usage of AI tools lately for my daily work and personal projects. I recently went through deeplearning's langchain course and was amazed about it. Normally, I work with .NET, so this project would be a great tool for me.
I'd be glad to contribute to the project in my free time, but would probably end up using this for my work-related development depending on licensing!
It's great to be here. Hope to see this community grow.

**Reply by HavenDV** [reply]
Hi @reinarrow. Great! Glad to see that the project is interesting to people. 
At the moment, I'm trying to consistently set aside a little bit of time each day to progressively develop the repository. In the near future, I want to implement a basic task, like QnA on PDF, so that the project begins to bring real benefits.

> using this for my work-related development depending on licensing!

The project is under the MIT license, and usually many employers allow you to use it in your work.

**Comment by manikanthvangala** [comment]
hi @HavenDV ,

Can i use this API in my production ? is it licensed ?

**Reply by HavenDV** [reply]
Hi. It's released under the MIT license, so yes. But this is still in a very early state of development and many scenarios are untested, so you need to be careful.

**Reply by manikanthvangala** [reply]
Thanks @HavenDV

**Reply by manikanthvangala** [reply]
hi @HavenDV 
I am trying to build my project with your library. It is throwing me below error

![Screenshot 2023-11-22 131551](https://github.com/tryAGI/LangChain/assets/67571591/9109f430-48e1-4c9d-aaab-22ec9a124839)


Could you help me why this so ? When i remove the library and build in release mode. Its working fine. But with the library in debug mode its working good

**Reply by TesAnti** [reply]
this may be an older version of visual studio. have you tried to update to the latest version?

**Reply by HavenDV** [reply]
There is also a small chance that something we use is not compatible with the .Net Framework, so also write which TargetFramework you are using.
But first of all, I would, of course, update Rider/Visual Studio to the latest stable version and use dotnet 8 sdk

**Reply by manikanthvangala** [reply]
hi @HavenDV / @TesAnti 

My current dotnet --version
**8.0.100**

```
<PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x64'">
    <OutputPath>bin\x64\Release\</OutputPath>
    <DefineConstants>TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <Optimize>true</Optimize>
    <NoWarn>;2008</NoWarn>
    <DebugType>pdbonly</DebugType>
    <PlatformTarget>x64</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
    <UseDotNetNativeToolchain>true</UseDotNetNativeToolchain>
    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
  </PropertyGroup>
```
  
```
  <PackageReference Include="Microsoft.AppCenter">
      <Version>4.5.3</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.AppCenter.Analytics">
      <Version>4.5.3</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.AppCenter.Crashes">
      <Version>4.5.3</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.Data.Sqlite">
      <Version>5.0.10</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.Graph">
      <Version>4.7.0</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.Net.Native.Compiler">
      <Version>2.2.3</Version>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
    <PackageReference Include="Microsoft.NETCore.UniversalWindowsPlatform">
      <Version>6.2.14</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.Toolkit.Uwp">
      <Version>7.1.2</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.Toolkit.Uwp.UI">
      <Version>7.1.2</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.Toolkit.Uwp.UI.Controls">
      <Version>7.1.2</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.Toolkit.Uwp.UI.Media">
      <Version>7.1.2</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.Win32.Registry">
      <Version>5.0.0</Version>
    </PackageReference>
<PackageReference Include="tryAGI.OpenAI">
      <Version>1.8.2</Version>
    </PackageReference>
```

Below is my VS info

Microsoft Visual Studio Community 2022
Version 17.8.1
VisualStudio.17.Release/17.8.1+34316.72
Microsoft .NET Framework
Version 4.8.09032

Installed Version: Community

Visual C++ 2022   00482-90000-00000-AA027
Microsoft Visual C++ 2022

ASP.NET and Web Tools   17.8.352.38654
ASP.NET and Web Tools

Azure App Service Tools v3.0.0   17.8.352.38654
Azure App Service Tools v3.0.0

C# Tools   4.8.0-7.23558.1+e091728607ca0fc9efca55ccfb3e59259c6b5a0a
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Extensibility Message Bus   1.4.39 (main@e8108eb)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Mono Debugging for Visual Studio   17.8.17 (957fbed)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   6.8.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

Razor (ASP.NET Core)   17.8.3.2353105+f0a326f3e15b2370371693a76fd379635387c74d
Provides languages services for ASP.NET Core Razor.

Test Adapter for Boost.Test   1.0
Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

Test Adapter for Google Test   1.0
Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

TypeScript Tools   17.0.20920.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   4.8.0-7.23558.1+e091728607ca0fc9efca55ccfb3e59259c6b5a0a
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   17.8.0-beta.23475.2+10f956e631a1efc0f7f5e49c626c494cd32b1f50
Microsoft Visual F# Tools

Visual Studio IntelliCode   2.2
AI-assisted development for Visual Studio.

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

VSPackage Extension   1.0
VSPackage Visual Studio Extension Detailed Info

Xamarin   17.8.0.155 (d17-8@7042beb)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   17.8.3.6 (remotes/origin/d17-8@eccf46a291)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin.Android SDK   13.2.2.0 (d17-5/45b0e14)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: d9a6e87
    Java.Interop: xamarin/java.interop/d17-5@149d70fe
    SQLite: xamarin/sqlite/3.40.1@68c69d8
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d17-5@ca1552d

![image](https://github.com/tryAGI/LangChain/assets/67571591/4cd4e12d-f4e2-44c3-a5f8-03e97647e9dd)

**Reply by HavenDV** [reply]
It was because UseDotNetNativeToolchain = true: https://stackoverflow.com/questions/36655045/internal-compiler-error-when-usedotnetnativetoolchain-is-true

**Comment by SiegDuch** [comment]
Hello everyone, when I first stumbled upon this project, (some weeks ago) I was looking to find a way to use LangChain's ability to make Retrieval Augmented Generation (RAG) but with c# code. I'm kind of new to everything around AI and learning about it. I'm currently using Vertex AI for a client but as far as I understand, everything could be connected easily. I'm also a programmer with some background and I'm always happy to help others.

**Reply by HavenDV** [reply]
Hello. While I would like to recommend this project, there are still many challenges to overcome and different scenarios to test before it can actually be used. Therefore, I think in your case using Vertex AI is preferable.
But, you can help us if you explore the possibility of adding integration with Vertex AI like in the original LangChain: https://python.langchain.com/docs/integrations/llms/google_vertex_ai_palm
